src/main/kotlin/uk/gov/gdx/datashare/queue/HmppsQueueFactory.kt

```diff
@@ -0,0 +1,140 @@
+    if (queueConfig.dlqName.isNotEmpty()) {
+      getOrDefaultBean("$queueId-sqs-dlq-client") {
+        createSqsDlqClient(queueId, queueConfig, hmppsSqsProperties)
+      }
+    } else null
+
+  private fun getOrDefaultSqsClient(queueId: String, queueConfig: HmppsSqsProperties.QueueConfig, hmppsSqsProperties: HmppsSqsProperties, sqsDlqClient: AmazonSQS?): AmazonSQS =
+    getOrDefaultBean("$queueId-sqs-client") {
+      createSqsClient(queueId, queueConfig, hmppsSqsProperties, sqsDlqClient)
+    }
+
+  private fun getOrDefaultHealthIndicator(hmppsQueue: HmppsQueue): HealthIndicator =
+    getOrDefaultBean("${hmppsQueue.id}-health") {
+      HmppsQueueHealth(hmppsQueue)
+    }
+
+  private fun createJmsListenerContainerFactory(hmppsQueue: HmppsQueue, hmppsSqsProperties: HmppsSqsProperties): HmppsQueueDestinationContainerFactory =
+    getOrDefaultBean("${hmppsQueue.id}-jms-listener-factory") {
+      HmppsQueueDestinationContainerFactory(hmppsQueue.id, createJmsListenerContainerFactory(hmppsQueue.sqsClient, hmppsSqsProperties))
+    }
+
+  @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
+  private inline fun <reified T> getOrDefaultBean(beanName: String, createDefaultBean: () -> T) =
+    runCatching { context.beanFactory.getBean(beanName) as T }
+      .getOrElse {
+        createDefaultBean().also { bean -> context.beanFactory.registerSingleton(beanName, bean) }
+      }
+
+  fun createSqsDlqClient(queueId: String, queueConfig: HmppsSqsProperties.QueueConfig, hmppsSqsProperties: HmppsSqsProperties): AmazonSQS =
+    with(hmppsSqsProperties) {
+      if (queueConfig.dlqName.isEmpty()) throw MissingDlqNameException()
+      when (provider) {
+        "aws" -> amazonSqsFactory.awsSqsDlqClient(queueId, queueConfig.dlqName, queueConfig.dlqAccessKeyId, queueConfig.dlqSecretAccessKey, region)
+        "localstack" ->
+          amazonSqsFactory.localStackSqsDlqClient(queueId, queueConfig.dlqName, localstackUrl, region)
+            .also { sqsDlqClient -> sqsDlqClient.createQueue(queueConfig.dlqName) }
+
+        else -> throw IllegalStateException("Unrecognised HMPPS SQS provider $provider")
+      }
+    }
+
+  fun createSqsClient(queueId: String, queueConfig: HmppsSqsProperties.QueueConfig, hmppsSqsProperties: HmppsSqsProperties, sqsDlqClient: AmazonSQS?) =
+    with(hmppsSqsProperties) {
+      when (provider) {
+        "aws" -> amazonSqsFactory.awsSqsClient(queueId, queueConfig.queueName, queueConfig.queueAccessKeyId, queueConfig.queueSecretAccessKey, region)
+        "localstack" ->
+          amazonSqsFactory.localStackSqsClient(queueId, queueConfig.queueName, localstackUrl, region)
+            .also { sqsClient -> createLocalStackQueue(sqsClient, sqsDlqClient, queueConfig.queueName, queueConfig.dlqName, queueConfig.dlqMaxReceiveCount) }
+
+        else -> throw IllegalStateException("Unrecognised HMPPS SQS provider $provider")
+      }
+    }
+
+  private fun createLocalStackQueue(
```

Comment:
If you want to be really pedantic, it's a little confusing that the top function here creates a Queue object, where as this actually creates an SQS queue